27180 css nesting (#28473)
mdn/content: files/en-us/web/css/css_nesting/using_css_nesting/index.md

---
title: Using CSS nesting
slug: Web/CSS/CSS_nesting/Using_CSS_nesting
page-type: guide
---

{{CSSRef}}

The [CSS Nesting](/en-US/docs/Web/CSS/CSS_nesting) module allows you to wite your stylesheets so that they are easier to read, more modular, and more maintainable. As you are not constantly repeating selectors the file size can also be reduced.

> CSS nesting is different from CSS preprocessors such as [Sass](https://sass-lang.com/) in that it is parsed by the browser rather than being pre-compiled by a CSS preprocessor.

Author: See the previous comment
https://github.com/mdn/content/pull/28473#discussion_r1301691958

```css
.notices {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  width: 90%;
  margin: auto;
}
```

Styles using nesting to create compound selectors. The top level selector creates the basic styles for an element with `class="notice"`, the the `&` nesting selector is used to create a compound selector for elements with either `class="notice warning"` or `class="notice success"`.

```css
.notice {
  width: 90%;
  justify-content: center;
  border-radius: 1rem;
  border: black solid 2px;
  background-color: #ffc107;
  color: black;
  padding: 1rem;
  h2:before {
    content: "ℹ︎ ";
  }
  &.warning {
    background-color: #d81b60;
    border-color: #d81b60;
    color: white;
    h2:before {
      content: "! ";
    }
  }
  &.success {
    background-color: #004d40;
    border-color: #004d40;
    color: white;
    h2:before {
      content: "✓ ";
    }
  }
}
```

##### Result

{{EmbedLiveSample('Nesting_and_compound_selectors','100%', '455')}}

## Appended nesting selector

The `&` nesting selector can also be appended to a nested selector which has the effect of reversing the context.

```css
.parent {
  /* .parent styles */
  .child & {

>     /* .child .parent styles */

Author: No here the example is for swapping the context of the selectors.
in this example:
```
<article class="card featured">
  <h2>Card 2</h2>
  <p>Lorem ipsum dolor, sit amet consectetur adipisicing elit.</p>
</article>
```
We can have a style for `<h2>`s
```
h2 {/* h2 styles */}  
```
Then with nesting we can target `<h2>`s that are children of a **featured** card by appending the `&` nesting selector

```
h2 {
  /* h2 styles */
  .featured & {/* .featured h2 styles */}
}  
```
I shall change the class names as `.parent` & `.child` are confusing the issue

>     /* .child .parent styles */

Author: I have changed the class names to prevent confusion and added more explantion of this.

In this example there are 3 cards, one of which is featured. The cards are all exactly the same except the featured card will have an alternative color for the heading. By appending the `&` nesting selector the style for the .featured .h2 can be nested in the style for the `h2`.

##### HTML

```html
<div class="wrapper">
  <article class="card">
    <h2>Card 1</h2>
    <p>Lorem ipsum dolor, sit amet consectetur adipisicing elit.</p>
  </article>
  <article class="card featured">
    <h2>Card 2</h2>
    <p>Lorem ipsum dolor, sit amet consectetur adipisicing elit.</p>
  </article>
  <article class="card">
    <h2>Card 3</h2>
    <p>Lorem ipsum dolor, sit amet consectetur adipisicing elit.</p>
  </article>
</div>
```

##### CSS

```css
.wrapper {
  display: flex;
  flex-direction: row;
  gap: 0.25rem;
  font-family: system-ui;
}
```

In the following CSS we are creating the styles for `.card`, `.card h2` and then in the `h2` styles we nest the `.featured` class with the `&` nesting selector appended which creates a style for `.card .featured h2`.

```css
.card {
  padding: 0.5rem;
  border: 1px solid black;
  border-radius: 0.5rem;
  & h2 {
    color: slateblue;

>     .featured & {

Reviewer: ```suggestion
    .featured & {  /* equivalent to `.featured .card` */
```

      color: tomato;
    }
  }
}
```

##### Result

{{EmbedLiveSample('Appending_nesting_selector','100%','250')}}

> ## Concatenation

Reviewer: ```suggestion
## Concatenation isn't possible
```

not sure about this edit, but as people skim and don't read, we should make this more obvious either with this edit or by using a `> **Warning:**`

Reviewer: oh, other idea, mabye just do  ` ```css example-bad`

Reviewer: other idea, add this as a sub section to invalid nested style rules below

Author: Moving it to the a sub section of invalid nested style rules would remove this from the **In this article** page menu.

I'll add `example-bad` and add a warning around:
_This is not possible in CSS nesting: …_

Author: added `## Concatenation is not possible`, a warning and added the `example-bad` flag to the code block

In CSS preprocessors such as [Sass](https://sass-lang.com/), it is possible to use nesting to join strings to create new classes. This is common in CSS methodologies such as [BEM](http://getbem.com/naming/).

```css
.component {
  &__child-element {
  }
}
/* In Sass this becomes */
.component__child-element {
}
```

> This is not possible in CSS nesting: without a [combinator](/en-US/docs/Learn/CSS/Building_blocks/Selectors/Combinators) the nested selector is treated as a [type selector](/en-US/docs/Web/CSS/Type_selectors). Allowing concatenation would break this.

Reviewer: ```suggestion
This is not possible in CSS nesting: when a [combinator](/en-US/docs/Learn/CSS/Building_blocks/Selectors/Combinators) is not used, the nested selector is treated as a [type selector](/en-US/docs/Web/CSS/Type_selectors). Allowing concatenation would break this.
```

> For [compound selectors](/en-US/docs/Web/CSS/CSS_selectors/Selector_structure#compound_selector) using `&Element` (a [type selector](/en-US/docs/Web/CSS/Type_selectors)) would make the CSS invalid. The type selector must come first in a compound selector, so it must be written as `Element&`.

Reviewer: ```suggestion
In [compound selectors](/en-US/docs/Web/CSS/CSS_selectors/Selector_structure#compound_selector), the type selector must come first. Writing `&Element` (a [type selector](/en-US/docs/Web/CSS/Type_selectors)) makes the CSS selector, and the entire selector block, invalid. As the type selector must come first, the compound selector must be written as `Element&`.
```

```css example-good
.my-class {
  element& {
  }
}

/* the browser parses this to become a compound selector */
.my-class {
}
element.my-class {
}
```

## Invalid nested style rules

If a nested CSS rule is invalid then all of the enclosed styles will be ignored. This does not affect the parent or preceding rules.

In the following example, there is an invalid selector (`%` is not a valid character for selectors). The rule that includes this selector is ignored, but subsequent valid rules are not.

```css example-bad
.parent {
  /* .parent styles these work fine */
  & %invalid {
    /* %invalid styles all of which are ignored */
  }
  & .valid {
    /* .parent .valid styles these work fine */
  }
}
```

> ### Example

Reviewer: this example can be dropped (as it is addressed above)

Author: Removed this example

##### HTML

```html
<div class="wrapper invalid">
  <article class="card">
    <h2>Card 1</h2>
    <p class="card__content">
      Lorem ipsum dolor, <a href="#">sit amet</a> consectetur adipisicing elit.
    </p>
  </article>
  <article class="card featured">
    <h2>Card 2</h2>
    <p class="card__content">
      Lorem ipsum dolor, <a href="#">sit amet</a> consectetur adipisicing elit.
    </p>
  </article>
  <article class="card">
    <h2>Card 3</h2>
    <p class="card__content">
      Lorem ipsum dolor, <a href="#">sit amet</a> consectetur adipisicing elit.
    </p>
  </article>
</div>
```

##### CSS

```css
.wrapper {
  display: flex;
  flex-direction: row;
  gap: 0.25rem;
  font-family: system-ui;
}
```

In the `__content` rule all of the enclosed styles and rules are ignored. This means that the content will not be italic and the links will not be styled. The rest of the rules all work as expected. To fix this we could change the `&__content` to show the full class name — `& card__content`.

```css
.card {
  padding: 0.5rem;
  border: 1px solid black;
  border-radius: 0.5rem;

>   &__content {
>     font-style: italic;
>     & a {
>       color: slateblue;
>       text-decoration-style: wavy;
>       .featured & {
>         color: tomato;
>       }
>     }
>   }
>   & h2 {
>     color: slateblue;

Reviewer: ```suggestion
  &__content { /* invalid so ignored */
    font-style: italic;
    & a { /* ignored */
      color: slateblue;
      text-decoration-style: wavy;
      .featured & { /* ignored */
        color: tomato;
      }
    }
  }
  & h2 { /* equivalent to `.card h2` */
    color: slateblue;
```

Author: Removed this example

>     .featured & {

Reviewer: ```suggestion
    .featured & { /* equivalent to `.featured .card h2` */
```

Author: Removed this example

      color: tomato;
    }
  }
}
```

##### Result

{{EmbedLiveSample('trying_to_concatenate_a_string_in_css','100%','200')}}

## See Also

> - [CSS Nesting](/en-US/docs/Web/CSS/CSS_nesting)

Reviewer: ```suggestion
- [CSS nesting](/en-US/docs/Web/CSS/CSS_nesting) module
```

- [`&` nesting selector](/en-US/docs/Web/CSS/Nesting_selector)

> - [Nesting @rules](/en-US/docs/Web/CSS/CSS_nesting/Nesting_at-rules)

Reviewer: ```suggestion
- [Nesting `@` at-rules](/en-US/docs/Web/CSS/CSS_nesting/Nesting_at-rules)
```

---
title: Using CSS nesting
slug: Web/CSS/CSS_nesting/Using_CSS_nesting
page-type: guide
---

{{CSSRef}}

The [CSS Nesting](/en-US/docs/Web/CSS/CSS_nesting) module allows you to wite your stylesheets so that they are easier to read, more modular and more maintainable. As you are not constantly repeating selectors the file size can also be drastically reduced too.

This guide show some of the different ways the nesting in CSS can be arranged.

> **Note:** The early version of the specification did not allow the nesting of [Type selectors](/en-US/docs/Web/CSS/Type_selectors) without the [`&` nesting selector](/en-US/docs/Web/CSS/Nesting_selector). This has now been updated and can be used without the `&` nesting selector. At the time of writing Firefox supports the new version of the specification while Chrome and Safari support the old version of the specification and must use the `&` nesting selector to work.

## Child selectors

> You can use CSS nesting to target selectors that are child selectors of a parent. This can be done with or without the [`&` nesting selector](/en-US/docs/Web/CSS/Nesting_selector).

Author: done

```css
/* Without nesting selector */
parent {
  /* parent styles */
  child {
    /* child of parent styles */
  }
}
/* Without nesting selector */
parent {
  /* parent styles */
  & child {
    /* child of parent styles */
  }
}

> /* both these become */

Author: done

parent {
  /* parent styles */
}
parent child {
  /* child of parent styles */
}
```

### Examples

In these examples the `<input>`s inside `<label>`s are being styled differently to the `<input>`s as a sibling of `<label>`s.

#### HTML

```html
<form>
  <label for="name"
    >Name:
    <input type="text" id="name" />
  </label>
  <label for="email">email:</label>
  <input type="text" id="email" />
</form>
```

#### Without `&` nesting selector

> ```html hidden

Author: Fixed this

    color: white;
    background-color: black;
  }
}
*/
```

##### Result

{{EmbedLiveSample('Nesting_the_sibling_combinator','100%','135')}}

## Compound selectors

When using [compound selectors](/en-US/docs/Web/CSS/CSS_selectors/Selector_structure#compound_selector) in nested CSS you **have** to use the `&` nesting selector. This is because the browser will automatically add whitespace between selectors that do not use the `&` nesting selector.

In order to target an element with `class="a b"` the `&` nesting selector is needed otherwise the whitespace will break the compound selector.

```css
.a {
  /* styles for element with class="a" */
  .b {
    /* styles for element with class="b" which is a descendants of class="a" */
  }
  &.b {
    /* styles for element with class="a b" */
  }
}
/* this becomes */
.a {
  /* styles for element with class="a" */
}
.a .b {
  /* styles for element with class="b" which is a descendants of class="a" */
}
.a.b {
  /* styles for element with class="a b" */
}
```

### Example

#### Nesting and compound selectors

In this example the `&` nesting selector is used to create compound selectors to style elements with multiple classes.

##### HTML

> ```html

Author: I have completely rewritten this example

<div class="wrapper">
  <div class="a c"></div>
  <div class="a">
    <div class="b c"></div>
    <div class="b"></div>
  </div>
</div>
```

##### CSS

Styles for the `.wrapper` to create a {{cssxref('CSS_flexible_box_layout', 'flexbox layout')}}.

```css
.wrapper,
.a {
  display: flex;
  flex-direction: row;
  gap: 1rem;
  padding: 0.5rem;
}
```

Styles using nesting to create compound selectors.

```css
.a {
  /* styles for element with class="a" */
  height: 50px;
  width: 100px;
  border-radius: 0.5rem;
  border: red dotted 2px;
  .b {
    /* styles for element with class="a" and descendant with class="b" */
    height: 40px;
    width: 40px;
    border: black solid 2px;
    &.c {
      /* styles for element with class="b c" */
      border: blue solid;
    }
  }
  &.c {
    /* styles for element with class="a c" */
    border: green solid;
  }
}
```

##### Result

{{EmbedLiveSample('Nesting_and_compound_selectors','100%')}}

## Appended nesting selector

> Using the `&` nesting selector before a nested selector applies the styles to the child selector when it is within the parent selector.

Author: Removed this

```css
.parent {
  /* .parent styles */
  & .child {
    /* .parent .child styles */
  }
}
```

The `&` nesting selector can also be appended to a nested selector which has the effect of reversing the context.

```css
.parent {
  /* .parent styles */
  .child & {
    /* .child .parent styles */
  }
}
```

### Example

> #### Appending nesting selector

Author: Thanks
